File: tasks/text-redact/python-typed/environment/src/redact.py
"""Find the regions a rule matches and mask them."""

from parse import LiteralRule, Span

Found = tuple[int, int]
Attributed = tuple[int, int, str]


def find_spans(rule: LiteralRule, text: str) -> list[Found]:
    """Scan for the literal one position at a time."""
    spans: list[Found] = []
    width = len(rule["value"])
    index = 0
    while index + width <= len(text):
        if text[index : index + width] == rule["value"]:
            spans.append((index, index + width))
        index += 1
    return spans


def merge_spans(spans: list[Attributed]) -> list[Span]:
    """Spans are reported in the order they were found."""
    return [
        {"start": start, "end": end, "rules": [identifier]}
        for start, end, identifier in spans
    ]


def apply_mask(text: str, spans: list[Span], mask: str) -> str:
    characters = list(text)
    for span in spans:
        for index in range(span["start"], span["end"]):
            characters[index] = mask
    return "".join(characters)

File: tasks/redact-spans/go/src/redact.go
package main

type found struct {
	Start int
	End   int
}

type attributed struct {
	Start int
	End   int
	ID    string
}

type span struct {
	Start int      `json:"start"`
	End   int      `json:"end"`
	Rules []string `json:"rules"`
}

// findSpans scans for the literal one position at a time.
func findSpans(item rule, text string) []found {
	spans := []found{}
	width := len(item.Value)
	for index := 0; index+width <= len(text); index++ {
		if text[index:index+width] == item.Value {
			spans = append(spans, found{Start: index, End: index + width})
		}
	}
	return spans
}

// mergeSpans reports spans in the order they were found.
func mergeSpans(spans []attributed) []span {
	merged := []span{}
	for _, item := range spans {
		merged = append(merged, span{
			Start: item.Start,
			End:   item.End,
			Rules: []string{item.ID},
		})
	}
	return merged
}

func applyMask(text string, spans []span, mask string) string {
	characters := []byte(text)
	for _, item := range spans {
		for index := item.Start; index < item.End; index++ {
			characters[index] = mask[0]
		}
	}
	return string(characters)
}
